prql-compiler/src/sql/gen_expr.rs

```diff
@@ -76,132 +76,230 @@ pub(super) fn translate_expr(expr: Expr, ctx: &mut Context) -> Result<sql_ast::E
                 else_result,
             }
         }
-        ExprKind::BuiltInFunction { .. } => {
-            let expr = match try_into_is_null(expr, ctx)? {
-                Ok(is_null) => return Ok(is_null),
-                Err(expr) => expr,
-            };
-
-            let expr = match try_into_between(expr, ctx)? {
-                Ok(between) => return Ok(between),
-                Err(expr) => expr,
-            };
-
-            let expr = match try_into_concat_function(expr, ctx)? {
-                Ok(between) => return Ok(between),
-                Err(expr) => expr,
-            };
-
-            let expr = match try_into_binary_op(expr, ctx)? {
-                Ok(bin_op) => return Ok(bin_op),
-                Err(expr) => expr,
-            };
-
-            let expr = match try_into_unary_op(expr, ctx)? {
-                Ok(un_op) => return Ok(un_op),
-                Err(expr) => expr,
-            };
-
-            super::std::translate_built_in(expr, ctx)?
-        }
+        ExprKind::BuiltInFunction { ref name, ref args } => match &name[..] {
+            "std.eq" | "std.ne" => {
+                if args.len() == 2 {
+                    let (a, b) = (&args[0], &args[1]);
+
+                    if a.kind == ExprKind::Literal(Literal::Null)
+                        || b.kind == ExprKind::Literal(Literal::Null)
+                    {
+                        process_null(name, args, ctx)?
+                    } else if let Some(op) = operator_from_name(name) {
+                        let (left, right) = (&args[0], &args[1]);
+                        translate_binary_operator(left, right, op, ctx)?
+                    } else {
+                        translate_built_in(expr, ctx)?
+                    }
+                } else {
+                    super::std::translate_built_in(expr, ctx)?
+                }
+            }
+            "std.neg" | "std.not" => {
+                if args.len() == 1 {
+                    process_unary(name, &args[0], ctx)?
+                } else {
+                    super::std::translate_built_in(expr, ctx)?
+                }
+            }
+            "std.concat" => process_concat(&expr, ctx)?,
+            _ => match try_into_between(expr.clone(), ctx)? {
+                Some(between_expr) => between_expr,
+                None => {
+                    if let Some(op) = operator_from_name(name) {
+                        if args.len() == 2 {
+                            let (left, right) = (&args[0], &args[1]);
+                            translate_binary_operator(left, right, op, ctx)?
+                        } else {
+                            super::std::translate_built_in(expr, ctx)?
+                        }
+                    } else {
+                        super::std::translate_built_in(expr, ctx)?
+                    }
```

Comment:
This whole part is being repeated. Could it be converted into a fall-through case of the match?

```
ExprKind::BuiltInFunction { ref name, ref args } => {

    // special cases
    match name.as_str() {
        "std.eq" | "std.ne" => {
            if ... {
                return process_null(...);
            }
        },
        "std.neg" | "std.not" => {
            if args.len() == 1 {
                return process_unary(name, &args[0], ctx)?
            }
        }
        _ => {}
    }
    
    ... the fall-trough case ...
}
```

Reply:
Good idea; done